src/daft-logical-plan/src/optimization/rules/reorder_joins/join_graph.rs

```diff
@@ -0,0 +1,774 @@
+            adj_list.entry(r_ptr).or_default().push(l_ptr);
+        }
+        let start_ptr = Arc::as_ptr(&self.edges[0].0.plan);
+        let mut seen = HashSet::new();
+        let mut stack = vec![start_ptr];
+
+        while let Some(current) = stack.pop() {
+            if seen.insert(current) {
+                // If this is a new node, add all its neighbors to the stack.
+                if let Some(neighbors) = adj_list.get(&current) {
+                    stack.extend(neighbors.iter().filter(|&&n| !seen.contains(&n)));
+                }
+            }
+        }
+        seen.len() == adj_list.len()
+    }
+
+    /// Test helper function that checks if the graph contains the given projection/filter expressions
+    /// in the given order.
+    pub(crate) fn contains_projections_and_filters(&self, to_check: Vec<&ExprRef>) -> bool {
+        let all_exprs: Vec<_> = self
+            .final_projections_and_filters
+            .iter()
+            .flat_map(|p| match p {
+                ProjectionOrFilter::Projection(projs) => projs.iter(),
+                ProjectionOrFilter::Filter(f) => std::slice::from_ref(f).iter(),
+            })
+            .collect();
+        let mut check_idx = 0;
+        for expr in all_exprs {
+            if expr == to_check[check_idx] {
+                check_idx += 1;
+                if check_idx == to_check.len() {
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    /// Helper function that loosely checks if a given edge (represented by a simple string)
+    /// exists in the current graph.
+    pub(crate) fn contains_edge(&self, edge_string: &str) -> bool {
+        for edge in &self.edges {
+            if edge.simple_repr() == edge_string {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// JoinGraphBuilder takes in a logical plan. On .build(), it returns a JoinGraph that represents the given logical plan.
+struct JoinGraphBuilder {
+    join_conds_to_resolve: Vec<(String, LogicalPlanRef, bool)>,
+    final_name_map: HashMap<String, ExprRef>,
+    edges: Vec<JoinEdge>,
+    final_projections_and_filters: Vec<ProjectionOrFilter>,
+}
+
+impl JoinGraphBuilder {
+    pub(crate) fn build(self) -> JoinGraph {
+        JoinGraph::new(self.edges, self.final_projections_and_filters)
+    }
+
+    pub(crate) fn from_logical_plan(plan: LogicalPlanRef) -> Self {
+        let output_schema = plan.schema();
+        // During join reordering, we might produce an output schema that differs from the initial output schema. For example,
+        // columns might be rearranged, or columns that were not originally selected might now be in the output schema.
+        // Hence, we take the original output schema and turn it into a projection that we should apply after all other join
+        // ordering projections have been applied.
+        let output_projection = output_schema
```

Comment:
usually you would run the compute during the `build` method.
this would allow you to do the following pattern with a builder
```
let graph = JoinGraphBuilder::from_logical_plan(plan)
  .with_some_option(true)
  .with_some_threshold(10)
  .build();
```